src/kOS/UserIO/TelnetWelcomeMenu.cs

```diff
@@ -25,12 +25,248 @@ public class TelnetWelcomeMenu : MonoBehaviour
         private readonly StringBuilder localMenuBuffer;
         private bool firstTime;
         private volatile bool forceMenuReprint;
-        
+        private MenuLevel currentLevel;
+
+        private abstract class MenuLevel
+        {
+            public virtual MenuLevel parentMenu { get { return new OuterMenu (); } }
+
+            public virtual List<String> helpText { get; } = new List<String> { };
+
+            public virtual List<List<String> > headers { get; } = new List<List<String> > {
+                new List<String> { "Menu", "" },
+                new List<String> { "Pick", "" },
+                new List<String> { "----", "----------------" }
+            };
+            public virtual String format { get; } = "{0,4} {1}";
+
+            public virtual void PrintMenu (TelnetWelcomeMenu parent)
+            {
+
+                List<String> displayChoices = new List<string> { };
+                int longestLength = 0;
+                foreach (List<String> hdrLine in headers) {
+                    String choice = String.Format (format, hdrLine.Cast<object> ().ToArray ());
+                    displayChoices.Add (choice);
+                    longestLength = Math.Max (choice.Length, longestLength);
+                }
+                int i = 1;
+                foreach (object[] bodyLine in this.MenuItems()) {
+                    string choice = String.Format (format, new object[] { "[" + i.ToString () + "]" }.Concat (bodyLine).ToArray ());
+                    displayChoices.Add (choice);
+                    longestLength = Math.Max (choice.Length, longestLength);
+                    i++;
+                }
+                foreach (string choice in displayChoices) {
+                    string choicePaddedToLongest = choice + new String (' ', (longestLength - choice.Length));
+                    parent.telnetServer.Write (parent.CenterPadded (choicePaddedToLongest, ' ') + (char)UnicodeCommand.STARTNEXTLINE);
+                }
+            }
+
+            public virtual List< object[] > MenuItems ()
+            { // Deliberately empty, so this is allowed not to be implemented.
+                return new List< object[] > ();
+            }
+
+            public virtual int ItemCount ()
+            {
+                return MenuItems ().Count;
+            }
+
+            public virtual bool ItemsChanged ()
+            {
+                return false;
+            }
+
+            public virtual void Selected (TelnetWelcomeMenu parent, int k)
+            {
+                return;
+            }
+        }
+
+        private class OuterMenu : MenuLevel
```

Comment:
To summarize, @jonored don't bother trying to break out the nested classes unless you've already worked at it.  @Dunbaratu will be doing final review, and if he's good with the nesting I don't have an issue having it here.